### ISSUE TITLE
Connect-HPERedFish Error

### ISSUE THREAD

**Issue by Johannes219** [HewlettPackard/PowerShell-ProLiant-SDK#74]
Hi,

i try to use the HPERedFishCMdlets for PowerShell (v7.2) on a Windows 2012 R2 System, but i got the following error when i try to connect to an HPE Proliant DL380 Gen10 Server (ILO 5):

$cred = Import-Clixml -Path "[path to credentials.xml]"  
Connect-HPERedfish -Address [ip-address] -Credential $cred -DisableCertificateAuthentication 

Error: InvalidOperation: The property 'Expect100Continue' cannot be found on this object. Verify that the property exists and can be set.

Any help is appreciated.

Regards,
Johannes

**Comment by ariael** [comment]
Hi
Do you have expected version .NET?
Best Regards
Peter Ari

**Comment by Johannes219** [comment]
Hey,

the last installed Version should be .NET Framework 4.8.

In Reference to 
https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed

the Release ID for Version 4.8 is similar to the entries in the Windows 2012 Registry:
![NET Framework](https://user-images.githubusercontent.com/96593725/147336832-a1c1aaca-6733-4a0f-8d57-d2b3da4ac5e6.png)

Best Regards,
Johannes

**Comment by Johannes219** [comment]
Small Update - i have found out, that the connect-hpeRedFish Command works correctly on the windows integrated PowerShell 5.1:
![ps5](https://user-images.githubusercontent.com/96593725/147404081-5e53c837-1b6c-4f03-a328-03497c5ff646.png)

Unfortunately i haven't found out the difference to PowerShell 7.2 until yet:
![ps7-error](https://user-images.githubusercontent.com/96593725/147404087-d7ee5110-270f-4e3a-9ce6-81a9529b038a.png)

However, i have one difference. The System has Internet Connection over an authenticating proxy. Although i tried to removing any customized proxy settings, in contrast to PowerShell 5.1, PowerShell 7 starts every new session with an definied proxy. So i must set "[system.net.webrequest]::DefaultWebProxy = new-object system.net.webproxy($null)" to remove the proxy settings. I have checked out every settings, but i haven't found out, where PowerShell is getting the proxy settings, Neither PowerShell Profile, Internet Proxy Settings, nor  envirenonment variables like http_proxy or https_proxy are set. Policy Settings are excluded- this proxy is not distributed over the system policies.

I don't know if this could affect the Expect100Continue Problem.

**Comment by Johannes219** [comment]
Hi,

is there no one who can help here?

Best regards
Johannes

**Comment by ariael** [comment]
Support from HP side is very very sad story...
You can try contact your HP dealer and push over this contact.
But sad story - this part of HP don't "talk" with other parts - businesses
and similar. And customers is on very very low level...
This was reason for very big problems in past with non compatible firmware
and this software in HP servers ....
Sorry for this bad news.
This product looks like good way for solve lots of problems, but this
unbelievable lacking support fully destroy this product.

My suggestion - buy servers from another company. I don't believe, than HP
have any real interest on better situation.
Any managers don't read this messages from customers.

Dne čt 30. 12. 2021 12:17 uživatel Johannes219 ***@***.***>
napsal:

**Comment by Johannes219** [comment]
@ariael It seems so..

i simply have deactivated the Expect100Continue now:

`$httpWebRequest.ServicePoint.Expect100Continue = $false`
 

But when i try to connect i become to followed error:

`Exception: Exception calling "GetResponse" with "0" argument(s): "The SSL connection could not be established, see inner exception."`
 

With "$Errors[0].Exception.InnerException | FL *" i get the followed error:

`InvalidOperation: Cannot index into a null array.`
 

I have found some solutions on https://stackoverflow.com/questions/34331206/ignore-ssl-warning-with-powershell-downloadstring to set ServerCertificateValidationCallback to $true, but it did not help:

`[System.Net.ServicePointManager]::ServerCertificateValidationCallback = {$true}`
 

Also, to get the line of this error, i have copied the CMDLet Functions in own script and renamed the function "Connect-HPERedfish2"
The Result of this Error is on Line 1723:

`[System.Net.WebResponse] $resp = $httpWebRequest.GetResponse()`
 

All commands to create variable $httpsWebRequest before the Line 1723 are listed below:

```
$wr = [System.Net.WebRequest]::Create($Uri)
$httpWebRequest = [System.Net.HttpWebRequest]$wr
$httpWebRequest.Method = $Method
$httpWebRequest.AutomaticDecompression = [System.Net.DecompressionMethods]::GZip
$httpWebRequest.Headers.Add('Odata-version','4.0')
$httpWebRequest.ServicePoint.Expect100Continue = $false
$httpWebRequest.Headers.Add('X-Auth-Token',$Session.'X-Auth-Token'
$httpWebRequest.ServerCertificateValidationCallback = {$true}
```
 

The $session variable is definied as followed:

```
$session = New-Object PSObject
$session|Add-Member -MemberType NoteProperty 'RootUri' $rootUri
$session|Add-Member -MemberType NoteProperty 'X-Auth-Token' $webResponse.Headers['X-Auth-Token']
$session|Add-Member -MemberType NoteProperty 'Location' $webResponse.Headers['Location']
$session|Add-Member -MemberType NoteProperty 'RootData' $rootData
$session|Add-Member -MemberType NoteProperty 'DisableExpect100Continue' $DisableExpect100Continue
```
 

Additionally, i have tried to connect to the Server with the Invoke-Webrequest:

`Invoke-WebRequest https://<IP-Address>/ -SkipCertificateCheck`
This was successfull.

Referring to the inner Exception, the $httpWebRequest.GetResponse() is $null. But why?

**Comment by Johannes219** [comment]
Hello,

i have a couple of updates - while i create my own script depending on the HPE ConnectCMDLet, i have found out, that the system proxy settings works against me.

With the line:
`[System.Net.WebResponse] $resp = $httpWebRequest.GetResponse()`

i got the following error:

```
 Line |
   72 |  [System.Net.WebResponse] $resp = $httpWebRequest.GetResponse()
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      | Exception calling "GetResponse" with "0" argument(s): "The proxy tunnel request to proxy 'http://<proxyUri>/' failed with status code '403'.""
```

Doesn't matter that the authentication fails - the connection should definitly not use this proxy.

So, i tried to reset the proxy with following line:

 `[System.Net.Webrequest]::DefaultWebProxy = New-object System.Net.Webproxy($null)`

This results in the root failure again:

```
 Line |
   69 |  $httpWebRequest.ServicePoint.Expect100Continue = -not($DisableExpect1 …
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      | The property 'Expect100Continue' cannot be found on this object. Verify that the property exists and can be set.
```

Then, i tried to configure the webproxy including bypass parameters in reference to https://stackoverflow.com/questions/28434480/creating-proxy-with-bypass-list-fails-if-it-contains

```
 $proxyUri = "http<proxy:port>"
 $snwr = [System.Net.Webrequest]::DefaultWebProxy = new-object System.Net.Webproxy($proxyUri, $true)
 $snwr.Credentials = Import-Clixml -Path "<path-to-credentials>"
 $snwr.BypassArrayList.Add("localhost")
 $snwr.BypassArrayList.Add("*.local")
 $snwr.BypassArrayList.Add("127.0.0.0/8")
 $snwr.BypassArrayList.Add("172.16.0.0/16")
 $snwr.BypassArrayList.Add("192.168.0.0/16")
 $snwr.BypassProxyOnLocal = $true
```

But the bypasslist does not work. The only thing that works is "localhost':

```
 $snwr.IsBypassed("https://localhost")                      >> returns true
 $snwr.IsBypassed("https://192.168.1.1")                  >> returns false
 $snwr.IsBypassed("https://website.local")                >> returns false
```

Additionally, i used the ip-addresses without apostrophes, but without success. Even the "192.168.* does not work.
There is annother property, called "BypassArrayList.AddRange" There, i don't know which value is valid, i become the followed error every time:

```
 Line |
   47 |  $snwr.BypassArrayList.AddRange(192.168.0.0/8)
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      | Cannot convert argument "c", with value: "0", for "AddRange" to type "System.Collections.ICollection": "Cannot convert the "0" value of type "System.Int32" to type "System.Collections.ICollection"."
```

In Conclusion - its not a basically error in powerShell 7, more an error with the proxy settings. I hope, when the proxy settings are correct / bypassed, the connection to the server will success.

How do I configure the proxy bypass settings to work properly?

Regards
Johannes

**Comment by Johannes219** [comment]
Hi,

The proxy settings are reset does not cause problems anymore.

The root cause for this problem are incompatibilities about the ServerCertificateValidationCallback property of ServicePointManager in PowerShell Core+.

The property is only supported in PowerShell 5.1, not in PowerShell Core and newer.

Links:
https://get-powershellblog.blogspot.com/2017/12/powershell-core-web-cmdlets-in-depth.html

https://github.com/PowerShell/PowerShell/issues/13597

[ServerCertificateValidationCallback causes problems with secure connections · Issue #13597 · PowerShell/PowerShell](https://ebookreading.net/view/book/EB9781789536669_423.html#)

[Capturing SSL errors](https://stackoverflow.com/questions/43810101/getting-unable-to-find-type-system-net-servicepointmanager-in-powershell-fo)

[macos - Getting "Unable to find type [System.Net.ServicePointManager]." in PowerShell for OS X - Stack Overflow](https://github.com/dotnet/runtime/issues/17154)
 

Is there a way to update the HPE RedFish CMDLets?

Regards

Johannes

**Comment by hp-powershell** [comment]
Hi ,

We don't have any scheduled release planned for Redfish cmdlets currently . We will let you know if management decides on a new release.

Thanks & Regards,
Powershell Team
